File: backend/services/categorizer.py
import os
import json
import re
from services.gemini_client import call_gemini_json
from services.file_parser import parse_first_page, get_chunks


def _storage(admin_id: int) -> str:
    return os.path.join("storage", str(admin_id))


def safe_name(name: str) -> str:
    return re.sub(r'[/\\:*?"<>|]', '_', name).strip()


def _normalize_tree(data: list) -> list:
    """구 형식(sub_categories)을 새 트리 형식(children)으로 변환"""
    result = []
    for item in data:
        if "sub_categories" in item:
            result.append({
                "name": item["name"],
                "children": [{"name": s, "children": []} for s in item.get("sub_categories", [])]
            })
        else:
            children = item.get("children", [])
            result.append({
                "name": item["name"],
                "children": _normalize_tree(children) if children else []
            })
    return result


def get_leaf_paths(tree: list, prefix: str = "") -> list:
    """트리에서 모든 리프 파일 경로 목록 반환 (상대경로/파일명.txt)"""
    paths = []
    for node in tree:
        name = safe_name(node["name"])
        path = (prefix + "/" + name) if prefix else name
        children = node.get("children", [])
        if children:
            paths.extend(get_leaf_paths(children, path))
        else:
            paths.append(path + ".txt")
    return paths


def propose_categories(uploaded_files: list) -> list:
    summaries = []
    for path in uploaded_files:
        if not os.path.exists(path):
            continue
        first_page = parse_first_page(path)
        summaries.append(f"[파일명: {os.path.basename(path)}]\n{first_page[:500]}")

    if not summaries:
        return []

    nl = "\n"
    prompt = f"""다음은 여러 문서들의 파일명과 첫 페이지 내용이야.
이 문서들을 분류하기 위한 디렉토리 트리 구조를 제안해줘.
내용이 단순하면 1~2단계, 복잡하면 3~4단계까지 만들어도 돼.
children이 비어있는 노드가 실제 파일(리프)이 되고, 나머지는 폴더야.

{nl.join(summaries)}

다음 JSON 형식으로만 응답해. 다른 텍스트는 절대 포함하지 마:
{{
  "tree": [
    {{
      "name": "폴더명",
      "children": [
        {{
          "name": "하위폴더 또는 파일명",
          "children": []
        }}
      ]
    }}
  ]
}}"""

    try:
        result = call_gemini_json(prompt)
        return result.get("tree", [])
    except Exception as e:
        print(f"Error proposing categories: {e}")
        return []


def save_categories(tree: list, admin_id: int):
    """트리 구조에 따라 폴더/파일 생성 — 항상 초기화"""
    storage = _storage(admin_id)
    os.makedirs(storage, exist_ok=True)

    with open(os.path.join(storage, "categories.json"), "w", encoding="utf-8") as f:
        json.dump({"tree": tree}, f, ensure_ascii=False, indent=2)

    def create_node(node, parent_path):
        name = safe_name(node["name"])
        children = node.get("children", [])
        if children:
            folder_path = os.path.join(parent_path, name)
            os.makedirs(folder_path, exist_ok=True)
            for child in children:
                create_node(child, folder_path)
        else:
            with open(os.path.join(parent_path, name + ".txt"), "w", encoding="utf-8") as f:
                pass

    for node in tree:
        create_node(node, storage)


def process_document(file_path: str, tree: list, admin_id: int):
    if not os.path.exists(file_path):
        return

    storage = _storage(admin_id)
    leaf_paths = get_leaf_paths(tree)
    if not leaf_paths:
        return

    chunks = get_chunks(file_path)
    source_name = os.path.basename(file_path)
    leaf_list = "\n".join([f"{i+1}. {p}" for i, p in enumerate(leaf_paths)])

    for chunk in chunks:
        if not chunk.strip():
            continue

        prompt = f"""다음 텍스트에서 아래 각 경로에 해당하는 내용을 추출해줘.
해당 내용이 없으면 빈 문자열로 남겨줘.
조항 번호, 항목 번호, 원본 구조를 그대로 유지해서 추출해.
설명이나 다른 텍스트는 절대 추가하지 마.

경로 목록:
{leaf_list}

[텍스트]
{chunk[:3000]}

JSON으로만 응답:
{{"extractions": [{{"path": "경로", "content": "추출내용"}}]}}"""

        try:
            result = call_gemini_json(prompt)
            for item in result.get("extractions", []):
                rel_path = item.get("path", "").strip()
                content = item.get("content", "").strip()

                if not content or content == "빈 문자열" or len(content) < 5:
                    continue
                if content.startswith("에러가 발생했습니다:"):
                    continue

                full_path = os.path.join(storage, rel_path.replace("/", os.sep))
                if os.path.exists(full_path):
                    with open(full_path, "a", encoding="utf-8") as f:
                        f.write(f"\n\n--- 출처: {source_name} ---\n")
                        f.write(content)
        except Exception as e:
            print(f"Error processing chunk from {source_name}: {e}")


def get_categories(admin_id: int) -> list:
    path = os.path.join(_storage(admin_id), "categories.json")
    if not os.path.exists(path):
        return []
    try:
        with open(path, "r", encoding="utf-8") as f:
            data = json.load(f)
        raw = data.get("tree", data.get("categories", []))
        return _normalize_tree(raw)
    except Exception as e:
        print(f"Error loading categories.json: {e}")
        return []

File: lib/categorizer.ts
import { callGeminiJson } from './gemini';
import { supabaseAdmin, STORAGE_BUCKET } from './supabase';
import { encodeStoragePath } from './storage-key';

export interface TreeNode {
  name: string;
  children?: TreeNode[];
}

const FILE_EXT_RE = /\.(txt|md|pdf|docx|doc|xlsx|csv|hwp)$/i;

export function safeName(name: string): string {
  return name.replace(FILE_EXT_RE, '').replace(/[/\\:*?"<>|]/g, '_').trim();
}

export function getLeafPaths(tree: TreeNode[], prefix = ''): string[] {
  const paths: string[] = [];
  for (const node of tree) {
    const name = safeName(node.name);
    const path = prefix ? `${prefix}/${name}` : name;
    if (node.children && node.children.length > 0) {
      paths.push(...getLeafPaths(node.children, path));
    } else {
      paths.push(`${path}.txt`);
    }
  }
  return paths;
}

export async function proposeCategories(
  filename: string,
  content: string
): Promise<TreeNode[]> {
  // Sample beginning + middle of document to avoid over-indexing on table of contents
  const len = content.length;
  const sample =
    content.slice(0, 1500) +
    (len > 3000 ? '\n...\n' + content.slice(Math.floor(len / 2), Math.floor(len / 2) + 1000) : '');

  const prompt = `다음 문서의 핵심 주제를 분석해서 간결한 카테고리 트리를 제안해줘.

파일명: ${filename}
내용 (일부):
${sample}

규칙:
- 이 문서 전체를 대표하는 최상위 폴더 1개만 만들 것
- 그 아래 핵심 주제별 파일(리프 노드)을 최대 6개까지만 만들 것
- 필요하면 중간 폴더를 1단계 추가할 수 있으나, 전체 깊이는 최대 3단계 이내
- 전체 리프 노드(파일) 수는 반드시 6개 이하
- 목차나 세부 절(節)을 그대로 나열하지 말고 대주제로 묶을 것
- 폴더명과 파일명은 한국어로, 내용을 잘 나타내도록
- 파일명에 확장자(.txt 등)를 절대 포함하지 말 것

JSON으로만 응답:
{"tree": [{"name": "폴더명", "children": [{"name": "파일명"}]}]}`;

  const result = await callGeminiJson(prompt);
  const raw = (result.tree as TreeNode[]) || [];
  return normalizeTree(raw);
}

export function normalizeTree(nodes: unknown[]): TreeNode[] {
  return nodes.map((n) => {
    const node = n as Record<string, unknown>;
    const children =
      (node.children as unknown[]) || (node.sub_categories as unknown[]) || [];
    const rawName = String(node.name || '');
    const cleanName = rawName.replace(FILE_EXT_RE, '').trim() || rawName;
    return {
      name: cleanName,
      ...(children.length > 0 ? { children: normalizeTree(children) } : {}),
    };
  });
}

export async function processDocument(
  adminId: number,
  _filename: string,
  chunks: string[],
  tree: TreeNode[],
  onProgress?: (msg: string) => void
): Promise<void> {
  const leafPaths = getLeafPaths(tree);

  for (let i = 0; i < chunks.length; i++) {
    const chunk = chunks[i];
    onProgress?.(`청크 ${i + 1}/${chunks.length} 처리 중...`);
    if (i > 0) await new Promise(r => setTimeout(r, 1000));

    const leafList = leafPaths.map((p, idx) => `${idx + 1}. ${p}`).join('\n');
    const prompt = `다음 문서 내용에서 각 파일 경로에 해당하는 내용을 추출해줘.

문서 (청크 ${i + 1}/${chunks.length}):
${chunk.slice(0, 3000)}

저장 경로 목록:
${leafList}

각 경로에 해당하는 내용만 추출해줘. 해당 없으면 빈 문자열.
JSON으로만 응답:
{"extractions": [{"index": 1, "content": "추출된 내용"}]}`;

    try {
      const result = await callGeminiJson(prompt);
      const extractions = (result.extractions as Array<{ index: number; content: string }>) || [];

      for (const extraction of extractions) {
        const leafPath = leafPaths[extraction.index - 1];
        if (!leafPath) continue;
        const content = extraction.content?.trim();
        if (!content || content.startsWith('에러가 발생했습니다:')) continue;

        const storagePath = `${adminId}/${encodeStoragePath(leafPath)}`;

        // Read existing content and append
        const { data: existing } = await supabaseAdmin.storage
          .from(STORAGE_BUCKET)
          .download(storagePath);

        let existingText = '';
        if (existing) {
          existingText = await existing.text();
        }

        const newContent = existingText
          ? `${existingText}\n\n${content}`
          : content;

        await supabaseAdmin.storage
          .from(STORAGE_BUCKET)
          .upload(storagePath, newContent, {
            contentType: 'text/plain; charset=utf-8',
            upsert: true,
          });
      }
    } catch (err) {
      console.error(`Chunk ${i + 1} processing error:`, err);
    }
  }

  onProgress?.('처리 완료');
}

export async function getCategories(adminId: number): Promise<TreeNode[]> {
  const path = `${adminId}/categories.json`;
  const { data, error } = await supabaseAdmin.storage
    .from(STORAGE_BUCKET)
    .download(path);

  if (error || !data) return [];

  try {
    const text = await data.text();
    const json = JSON.parse(text);
    const raw = json.tree || json.categories || [];
    return normalizeTree(raw);
  } catch {
    return [];
  }
}

export async function saveCategories(
  adminId: number,
  tree: TreeNode[]
): Promise<void> {
  const path = `${adminId}/categories.json`;
  const content = JSON.stringify({ tree }, null, 2);
  const { error } = await supabaseAdmin.storage
    .from(STORAGE_BUCKET)
    .upload(path, content, {
      contentType: 'application/json',
      upsert: true,
    });
  if (error) throw new Error(`Failed to save categories: ${error.message}`);
}
